### PR TITLE
Added support for Activation layers over tensor3 inputs

### PR DESCRIPTION
**Pull request by patyork**
Added a check in the Activation layer for a tensor3 input. If so, the tensor is reshaped into a matrix.

Assumed shape transformation: ```(nb_samples, time, nb_outputs)``` --> ```(nb_samples * time, nb_outputs)```

The activation (including softmax) can then be done correctly.

### CODE REVIEW

**Comment by fchollet** [comment]
As far as I can tell, the only activation function that wouldn't work out of the box on a tensor3 is Softmax (http://deeplearning.net/software/theano/library/tensor/nnet/nnet.html#tensor.nnet.softmax).

So I wonder, wouldn't it make more sense to add support for tensor3 inputs in the activations.softmax function instead? It seems cleaner to keep the Activation layer shape-agnostic, and it spares us a if branching when shape info is not relevant (which is most of the time). What do you think? Pros/cons?

**Comment by patyork** [comment]
I had the exact same line of thoughts, and below are what I came up with:

* Softmax is currently the only activation that I know of that won't support Softmax (it's actually the reason for this patch).
  * **Pro**: In this regard, putting it in the Softmax makes sense.
  * Con: But there's a chance that this may change, although a new activation that works across multiple outputs seems unlikely.
* Leaving it up to the user
  * **Con**: I believe, in the Theano documentation, that it states that the way that tensor3 should be read is: (batch, time, value). I may be wrong on this, I will look for a reference. If this is the case, we should probably follow this. I can't think of an instance where an tesnor3 would stand for something else..?
  * **Con**: The Reshape layer currently does not allow a changing of the size of the first dimension
https://github.com/fchollet/keras/blob/master/keras/layers/core.pyL80 . To allow this, I think we'd need to be able to reference the input shape, and somehow create a process through which the user can reference this input shape. This may be a _**big**_ problem on the last batch, where the actual number of samples != batch_size.

The cons outweigh the pros heavily in leaving it up to the user, and the Theano convention made me think that if a tensor3 is involved it follows that shape.

**Comment by patyork** [comment]
Or, as a thought, we might throw this in as an advanced_activation layer  https://github.com/fchollet/keras/blob/master/keras/layers/advanced_activations.py

My thinking is that the tensor3 requirement may only be for RNNs of different flavors, which need to utilize 3 dimensions for both correctness and speed.

Pros and Cons appreciated on this as well.

**Comment by fchollet** [comment]
From an architecture standpoint, my conclusion is that the best practice for a user wanting to use softmax on a 3D layer would be to do so with a Reshape layer (the first dimension is the number of samples, and therefore should never be touched).

**Comment by patyork** [comment]
This can't be done via a Reshape layer alone.

Ex: The RNN output is a tensor3 of (batch, time, value). Reshaping with a reshape layer gives an output shape of (batch, time*value). The Softmax will therefore not be applied over the correct values, and if we reshape back to (batch, time, value), the sum of the X[0][0][:] will != 1.0

**Comment by fchollet** [comment]
Is the idea to apply Softmax to every sample at every time step independently?

**Comment by patyork** [comment]
That's the only way in which it makes sense. For any dimensional data, the activation should be over the last dimension. In the case of softmax, there is some context that has to be retained. The last dimension cannot be reshaped.

**Comment by patyork** [comment]
Another alternative is a new Softmax3d activation function.

**Comment by fchollet** [comment]
That sounds like a good, pragmatic solution to me. With a totally explicit name, like TimeDistributedSoftmax.

**Comment by patyork** [comment]
If you'd like, I will close this request, and open a new one with the new activation.

I'm not sure on the name though, but we can called it TimeDistributedSoftmax until something better is offered.

**Comment by fchollet** [comment]
Ok, let's do this : ) 
I'm not sure on the name either, ideally what the activation does would be completely clear from the class name.
